### PR TITLE
Support zstd decompression for external dependencies

### PR DESCRIPTION
**Pull request by glukasiknuro**
Bug: [Add support for .tar.zst to http_archive](https://github.com/bazelbuild/bazel/issues/10342)

Handling of jni headers and loading system library is very similar to one already in bazel for native os library and possibly can be unified at some point.

Checked repo with big external deps, and noticed following improvements when checking bazel tracing profile:

```
size (tar.gz) |  extracting tar.gz  | extracting tar.zst
--------------------------------------------------------
    773M      |       13 s          |       3.2 s
    421M      |       13 s          |         5 s
    346M      |      9.2 s          |       3.3 s
    261M      |      7.5 s          |         3 s
```

One thing to check is whether bazel distributes binaries compiled with `-c opt` - personally tested with version compiled with this option.

### CODE REVIEW

**Comment by gjasny** [comment]
@googlebot I consent

**Comment by glukasiknuro** [comment]
@meisterT could you review this PR or suggest how to help with proceeding with this change?

**Comment by meisterT** [comment]
Sorry for the delay, I was OOO. 

What is the cost in terms of Bazel binary size with this change?

cc @meteorcloudy for the potential additional dependency

**Comment by glukasiknuro** [comment]
Tested on linux with `build //src:bazel --compilation_mode=opt` for file `bazel-bin/src/bazel`:

Diff: +497,142 (==+485kb, ==+1.06%)

Before (commit: ca6d7ac686108d5cdda6040409877d356ae627df) - 46,934,374  
After - 47,431,516

**Comment by meisterT** [comment]
That seems an acceptable cost to me. @meteorcloudy does this throw a wrench in your current Debian effort?

File: src/main/java/com/google/devtools/build/lib/bazel/repository/TarZstdFunction.java
Comment (meisterT) [inline src/main/java/com/google/devtools/build/lib/bazel/repository/TarZstdFunction.java:1]: Nit: we have 2020

File: src/test/shell/integration/minimal_jdk_test.sh
Comment (meisterT) [inline src/test/shell/integration/minimal_jdk_test.sh:51]: How much is it exactly after your change?

**Comment by meteorcloudy** [comment]
Yes, the zstd JNI library doesn't seem to be packaged for Debian yet. So we'll need to package that when we update the Bazel Debian package in the future.

/cc @olekw

File: src/test/shell/integration/minimal_jdk_test.sh
Comment (glukasiknuro) [reply src/test/shell/integration/minimal_jdk_test.sh:51]: It was failing on darwin: https://buildkite.com/bazel/bazel-bazel-github-presubmit/builds/6710#fec2a5d4-7cf4-47a6-bc19-215c8fd965b0 : 

`test_size_less_than_275MB FAILED: Size of install_base is 281984 kB, expected it to be less than 281600 kB.`

I may later try to check after/before on Mac.

**Comment by philwo** [comment]
What's the status of this PR?

File: src/test/shell/integration/minimal_jdk_test.sh
Comment (meisterT) [reply src/test/shell/integration/minimal_jdk_test.sh:51]: If you sync now, this change did already happen

**Comment by meisterT** [comment]
This looks okay to me, but I want to get @meteorcloudy's LGTM as well. If we agree to merge it, we need to split it into two separate PRs though (one only touching third_party, one not touching third_party).

File: src/test/shell/integration/minimal_jdk_test.sh
Comment (glukasiknuro) [reply src/test/shell/integration/minimal_jdk_test.sh]: Indeed there was commit https://github.com/bazelbuild/bazel/commit/2b028e48947e9484a6a7e1d505daffdb390b5174
Merged this change with current master.

**Review by meteorcloudy — APPROVED** [review]

**Comment by meteorcloudy** [comment]
Looks okay to me, too. But it will definitely add more work for packaging future version of Bazel for Debian.
/cc @olekw How difficult do you think to package https://github.com/luben/zstd-jni for Debian?

**Comment by gjasny** [comment]
Thank you for picking up the idea and turn it into production-level quality code! 🥇

**Comment by olekw** [comment]
> /cc @olekw How difficult do you think to package https://github.com/luben/zstd-jni for Debian?

Not very difficult. I discussed with the rest of the Debian Java Team a couple weeks ago and since no one was planning to work on it... I guess that means that I'm working on it. ;) The rest of the team does think that the package has broader value in Debian though so they will happily support it in the future. This is one of the things on my to-do list that I'm going to start chipping away at this week or next.

**Comment by meteorcloudy** [comment]
@olekw Thanks for the clarification!
@meisterT Looks like this PR is ready to be merged.

**Comment by philwo** [comment]
Could we replace the checked-in code with an `http_archive`? That would make it easier to merge and understand whether there were any changes made to the upstream sources.

**Comment by meteorcloudy** [comment]
> Could we replace the checked-in code with an http_archive? That would make it easier to merge and understand whether there were any changes made to the upstream sources.

That's a good point! 
@glukasiknuro Can you do that? An example change to add dependencies as http_archive is https://github.com/bazelbuild/bazel/commit/e2f11d7369672263ef126f87085fec0646c0c763

**Comment by glukasiknuro** [comment]
Changed to `http_archive`. Some points:

- to load native library needed to patch Native.java class in zstd-jni. Handling of native libraries changed in bazel as well (https://github.com/bazelbuild/bazel/commit/a2a2f9d5f6b), there maybe better way of implementing it in zstd-jni.
- needed to patch zstd-jni to properly compile under windows (MSVC). I've created PR to zstd-jni with similar change, but it waits for review - https://github.com/luben/zstd-jni/pull/140
- added integration test (note it's disabled on Windows as other tests in external integration dependency test file)

**Comment by meteorcloudy** [comment]
@glukasiknuro Thanks!

> - needed to patch zstd-jni to properly compile under windows (MSVC). I've created PR to zstd-jni with similar change, but it waits for review - luben/zstd-jni#140
> - added integration test (note it's disabled on Windows as other tests in external integration dependency test file)

Have you verified this PR on Windows in any way? If not I can help do that.

We cannot merge a PR that contains changes inside and outside of third_party directory at the same time. We'll have to split this PR into smaller ones:
- (third_party) Add the patch file for zstd
- (main) Add http_archive for zstd in WORKSPACE
- (third_party) update apache_commons_compress version and third_party/BUILD
- (main) Actual implementation and test

**Comment by glukasiknuro** [comment]
> @glukasiknuro Thanks!
> 
> > * needed to patch zstd-jni to properly compile under windows (MSVC). I've created PR to zstd-jni with similar change, but it waits for review - [luben/zstd-jni#140](https://github.com/luben/zstd-jni/pull/140)
> > * added integration test (note it's disabled on Windows as other tests in external integration dependency test file)
> 
> Have you verified this PR on Windows in any way? If not I can help do that.

Other than unit test no. If you could help with testing that would be awesome (I'll retest this PR on our large codebase after recent changes, but this is linux-only). 

> 
> We cannot merge a PR that contains changes inside and outside of third_party directory at the same time. We'll have to split this PR into smaller ones:
> 
> * (third_party) Add the patch file for zstd
> * (main) Add http_archive for zstd in WORKSPACE
> * (third_party) update apache_commons_compress version and third_party/BUILD
> * (main) Actual implementation and test

OK, let me prepare those four PRs in a moment (if that's the process). Let me update also zstd-jni to newer version that seems to have some small fixes.

**Comment by glukasiknuro** [comment]
@meteorcloudy To test it on windows you may use https://github.com/glukasiknuro/example-bazel-using-zstd-http-archive - it download an .tar.zst archive and allows to run some test.

**Comment by meteorcloudy** [comment]
It is failing on Windows with the following error
```
Internal error thrown during build. Printing stack trace: java.lang.ExceptionInInitializerError
        at org.apache.commons.compress.compressors.zstandard.ZstdCompressorInputStream.<init>(ZstdCompressorInputStream.java:43)
        at com.google.devtools.build.lib.bazel.repository.TarZstdFunction.getDecompressorStream(TarZstdFunction.java:39)
        at com.google.devtools.build.lib.bazel.repository.CompressedTarFunction.decompress(CompressedTarFunction.java:51)
        at com.google.devtools.build.lib.bazel.repository.DecompressorValue.decompress(DecompressorValue.java:126)
        at com.google.devtools.build.lib.bazel.repository.starlark.StarlarkRepositoryContext.downloadAndExtract(StarlarkRepositoryContext.java:916)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
        at java.base/java.lang.reflect.Method.invoke(Unknown Source)
        at net.starlark.java.eval.MethodDescriptor.call(MethodDescriptor.java:135)
        at net.starlark.java.eval.BuiltinCallable.fastcall(BuiltinCallable.java:72)
        at net.starlark.java.eval.Starlark.fastcall(Starlark.java:508)
        at net.starlark.java.eval.Eval.evalCall(Eval.java:628)
        at net.starlark.java.eval.Eval.eval(Eval.java:448)
        at net.starlark.java.eval.Eval.execAssignment(Eval.java:108)
        at net.starlark.java.eval.Eval.exec(Eval.java:260)
        at net.starlark.java.eval.Eval.execStatements(Eval.java:80)
        at net.starlark.java.eval.Eval.execFunctionBody(Eval.java:64)
        at net.starlark.java.eval.StarlarkFunction.fastcall(StarlarkFunction.java:155)
        at net.starlark.java.eval.Starlark.fastcall(Starlark.java:508)
        at net.starlark.java.eval.Starlark.call(Starlark.java:475)
        at com.google.devtools.build.lib.bazel.repository.starlark.StarlarkRepositoryFunction.fetch(StarlarkRepositoryFunction.java:210)
        at com.google.devtools.build.lib.rules.repository.RepositoryDelegatorFunction.fetchRepository(RepositoryDelegatorFunction.java:349)
        at com.google.devtools.build.lib.rules.repository.RepositoryDelegatorFunction.compute(RepositoryDelegatorFunction.java:276)
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:438)
        at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:398)
        at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: java.lang.RuntimeException: Unable to locate runfiles after failing to load zstd-jni from system. Giving up.
        at com.github.luben.zstd.util.Native.loadFromRunfiles(Native.java:62)
        at com.github.luben.zstd.util.Native.load(Native.java:53)
        at com.github.luben.zstd.ZstdInputStream.<clinit>(ZstdInputStream.java:21)
        ... 32 more
Caused by: java.io.IOException: Cannot find runfiles: $RUNFILES_DIR and $JAVA_RUNFILES are both unset or empty
        at com.google.devtools.build.runfiles.Runfiles.getRunfilesDir(Runfiles.java:180)
        at com.google.devtools.build.runfiles.Runfiles.create(Runfiles.java:114)
        at com.google.devtools.build.runfiles.Runfiles.create(Runfiles.java:79)
        at com.github.luben.zstd.util.Native.loadFromRunfiles(Native.java:60)
```

File: third_party/zstd-jni/zstd-jni_1.4.5-6.patch
Comment (meteorcloudy) [inline third_party/zstd-jni/zstd-jni_1.4.5-6.patch:125]: Why do you use the Runfiles library to find the libzstd-jni.dll? It is shipped within the Bazel binary, right?

Comment (meteorcloudy) [reply third_party/zstd-jni/zstd-jni_1.4.5-6.patch:125]: This is how we load JNI library in Bazel:
https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/jni/JniLoader.java;l=54?q=%22java.library.path%22&ss=bazel

I think here: `System.loadLibrary("libzstd-jni");` should just work.

Comment (glukasiknuro) [reply third_party/zstd-jni/zstd-jni_1.4.5-6.patch:125]: This code was based on the way Bazel did it before https://github.com/bazelbuild/bazel/commit/a2a2f9d5f6b that was commited 9 days ago. I'll try to unify it with current version (hopefully no need to create files), and test it under windows (setting up VM right now...) hopefully early next week.

**Comment by glukasiknuro** [comment]
PTAL

Fixed windows support - System.loadLibrary under windows does not add "lib" in from of shared library name.
Loading from Runfiles is needed in tests.

Also upgraded zstd-jni to 1.4.5-11 which contains fixes for pointer arithmetic which allows the patch to be smaller (https://github.com/luben/zstd-jni/pull/140). Would love to remove the patch for Native.java at all, but need to figure out how to do it so that zstd-jni will work as well.

Propagated changes to https://github.com/bazelbuild/bazel/pull/12274, and tested under windows and in our internal codebase (@meteorcloudy could you double check it works for you as well?)

**Comment by philwo** [comment]
@glukasiknuro I'm importing this now. I will have to split it up into a couple of commits due to the changes to //third_party, don't be surprised. :)

**Comment by philwo** [comment]
@glukasiknuro Two out of three parts are merged, I prepared a code review internally for the remaining changes - unfortunately I noticed that internally we appear to have a version of Apache Commons Compress without zstd support, so this doesn't compile when I build it in Google's monorepo. 😞 I'll have to figure out how to either upgrade our version of that library or change things so that our internal version of Bazel builds without the TarZstdFunction part (essentially disabling this feature just for our internal version).

**Comment by philwo** [comment]
@glukasiknuro Would it be an option to use https://github.com/airlift/aircompressor as an alternative to the JNI version?

**Comment by glukasiknuro** [comment]
> @glukasiknuro Would it be an option to use https://github.com/airlift/aircompressor as an alternative to the JNI version?

Using aircompressor would be nice b/c it's in pure java.

Although looks like aircompressor does not support streaming (i.e. whole file needs to be decompressed at once) and it does not support bigendian machines because of UNSAFE usage (this maybe smaller issue). But let me know if you think those are non-blockers, i may try to use aircompressor.

BTW tangential effort is supporting zstd to compress blobs in remote cache (https://github.com/bazelbuild/remote-apis/pull/168), I believe having zstd in some form would be needed for this effort as well. Looks like aircompressor implements only default zstd compression level, not sure how important is that though (@mostynb ?)

**Comment by mostynb** [comment]
> BTW tangential effort is supporting zstd to compress blobs in remote cache ([bazelbuild/remote-apis#168](https://github.com/bazelbuild/remote-apis/pull/168)), I believe having zstd in some form would be needed for this effort as well. Looks like aircompressor implements only default zstd compression level, but this can be probably OK (@mostynb ?)

The REAPI compressed-blob bytestream proposal leaves the compression level up to the implementations, I don't think using a library that only supports the default compression level would cause any interoperability issues. But you probably want to at least experiment with the default level and the "fastest" level before committing to a particular library.

**Comment by philwo** [comment]
> it does not support bigendian machines because of UNSAFE usage

That's unfortunately a blocker, as we don't want to break our s390x support. :/

I'm looking into merging the current version for now.

**Comment by philwo** [comment]
I'm sorry - I'm rolling back the partially merged PR before cutting Bazel 4.0 LTS, as we can't merge the remaining part in time and are worried that the newly added patched JNI dependency is too risky and will make it hard for distributions like Debian to accept.

Let's revisit this when we find a suitable pure Java version of Zstd.

**Comment by aignas** [comment]
Hello all, what would be required to have this feature in `bazel` given that `zstd` compression/decompression feature was landed for the grpc remote cache functionality?

**Comment by meteorcloudy** [comment]
@glukasiknuro Can you try to rebase with Bazel@HEAD and see if you can already use the existing zstd-jni library?

**Comment by meteorcloudy** [comment]
Closing due to https://github.com/bazelbuild/bazel/pull/15026

**Pull request closed without merging**
